[RFC] Crossover Filter

sebcarlucci opened the pull request:
Main Changes:
- Added crossover component
- Split up into functions the iir processing function. (if they are okay with this change, then maybe do it in a separate PR).

The configuration blob works very similarly to the equalizer, with the addition to an assign sink array.
The assign sink array, is used by the copy() function to figure out which buffer gets which frequency band.

sebcarlucci commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+	}
+}
+
+static int crossover_init_coef(struct comp_data *cd, int nch)
```
this function is a monster rn, do you think it should be split up into smaller chunks?

cujomalainey replied:
its definitely not the prettiest but it doesn't look to bad to me

sebcarlucci commented on src/audio/crossover/crossover_generic.c:
```diff
@@ -0,0 +1,157 @@
+
+	/* Cascade two biquads with same coefficients in series */
+	z = iir_process_biquad(in, lr4->coef, lr4->delay);
+	return iir_process_biquad(z, lr4->coef, &lr4->delay[2]);
```
modularizing the processing for the biquads also helps in reusing the same coefficient array.

sebcarlucci commented on src/include/kernel/abi.h:
```diff
@@ -30,7 +30,7 @@
 #define SOF_ABI_MAJOR 3
 #define SOF_ABI_MINOR 15
-#define SOF_ABI_PATCH 0
+#define SOF_ABI_PATCH 1
```
just added that to keep checkpatch quiet

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+#include <sof/audio/crossover/crossover.h>
+#include <sof/audio/eq_iir/iir.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
```
what is this used for?

sebcarlucci replied:
Its used for assert() calls

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+DECLARE_SOF_UUID("crossover", crossover_uuid, 0x948c9ad1, 0x806a, 0x4131,
+		 0xad, 0x6c, 0xb2, 0xbd, 0xa9, 0xe3, 0x5a, 0x9f);
+
+/* FOR TESTING ONLY - REMOVE THIS !!!!!!!! */
```
I would move this to a testing headerfile so you don't forget it :)

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+{
+	rfree(lr4->delay);
+
+	lr4->coef = NULL;
```
no free for the coef?

sebcarlucci replied:
The coefs are just pointer to memory in the config struct, so no need to free them

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+{
+	int i;
+
+	for (i = 0; i < 3; i++) {
```
magic number 3

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+				uint32_t pipe_id)
+{
+	int i;
+	int32_t *assign_sink = &config->data[0];
```
why not just config->data?

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+				  struct comp_buffer **sinks)
+{
+	struct comp_buffer *sink;
+	struct list_item *clist;
```
what is clist?

sebcarlucci replied:
from my understanding, clist is just an iterator used to go through the sink list

cujomalainey replied:
if its iterating over compnonents it would be nice to have a more verbose name

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+	 * TODO Have the check somehow
+	 */
+	if (bs > 0) {
+		comp_cl_err(&comp_crossover, "crossover_new(), NOT IMPLEMENTED YET");
```
i think this needs fixing?

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+			    struct sof_ipc_stream_params *params)
+{
+	comp_info(dev, "crossover_params()");
+
```
is there a TODO here?

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+		comp_info(dev, "crossover_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+
+		/* Copy back to user space */
+		if (cd->config) {
```
I'm seriously thinking there should be some helpers written for this boiler plate code. Hmm maybe I will write a PR, just a personal thought.

cujomalainey reviewed (changes requested):
i got up to the copy command, I will do the rest later tonight

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+	struct comp_buffer *sinks[CROSSOVER_MAX_STREAMS] = { NULL };
+	int i;
+	uint32_t num_sinks = 0;
+	uint32_t frames = -1;
```
you are assigning a signed numbed to an unsigned type

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+	uint32_t num_sinks = 0;
+	uint32_t frames = -1;
+	uint32_t source_bytes;
+	uint32_t avail;
```
stack these on the same line if you are not setting their values.

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+		crossover_free_config(&cd->config);
+		cd->config = cd->config_new;
+		cd->config_new = NULL;
+		/* TODO update coefficients with new config */
```
outstanding TODO?

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+	buffer_lock(source, &flags);
+
+	/* check if source is active */
+	if (source->source->state != dev->state) {
```
How far does this go? Could we potentially be the first element and hit a null dereference? This seems a bit weird.

sebcarlucci replied:
It would be an issue if we can have pipelines with no PCM or DAI? 
normally we have:
PCM -> B0 -> Crossover -> B1 -> DAI.

Could we potentially have this?:
Crossover -> B0 -> DAI

Maybe that could be a possibility if we were to "concatenate" pipelines together. An example would be sof-cml-demux-rt5682.m4.

cujomalainey replied:
sounds good to me, we will see how upstream likes it

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+		buffer_lock(sinks[i], &flags);
+		avail = audio_stream_avail_frames(&source->stream,
+						  &sinks[i]->stream);
+		frames = MIN(frames, avail);
```
how will this work in the case of an XRUN? This seems like we would block on it.

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+
+	for (i = 0; i < CROSSOVER_MAX_STREAMS; i++) {
+		if (!sinks[i])
+			continue;
```
can't this be break?

sebcarlucci replied:
Maybe the user did not assign the sinks correctly in the config blob, and therefore there could be a gap in the array

cujomalainey replied:
should we be allowing that to happen? Wouldnt that be an invalid config?

cujomalainey replied:
any update on this?

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+	/* update components */
+	for (i = 0; i < CROSSOVER_MAX_STREAMS; i++) {
+		if (!sinks[i])
+			continue;
```
ditto as comment above.

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+		return PPL_STATUS_PATH_STOP;
+
+	/* Crossover has a variable number of sinks. Assume that sink buffers
+	 * have same frame_fmt and size.
```
I would assert this assumption, see src comp for example

cujomalainey commented on src/audio/crossover/crossover_generic.c:
```diff
@@ -0,0 +1,157 @@
+// static void crossover_generic(struct crossover_state *state,
+//                               int32_t x, int32_t y[], int32_t num_sinks)
+// {
+//         /* PASS THROUGH FOR NOW */
```
leftover?

cujomalainey commented on src/audio/crossover/crossover_generic.c:
```diff
@@ -0,0 +1,157 @@
+		case 4:
+			crossover_generic_lr4_split(&state->lowpass[1],
+						    &state->highpass[1],
+						    z1 << 16, (int *)y1, &tmp);
```
magic number 16

cujomalainey commented on src/audio/crossover/crossover_generic.c:
```diff
@@ -0,0 +1,157 @@
+					    &state->highpass[0],
+					    *x << 16, (int *)&z1, (int *)&z2);
+
+		// TESTING FOR NOW
```
???

sebcarlucci replied:
I have set up this function right now to output to all of the channels of one sink. I still need to work on routing the output to their respective sink buffers.

cujomalainey commented on src/audio/eq_iir/iir_generic.c:
```diff
@@ -44,11 +44,44 @@
 /* 32 bit data, 32 bit coefficients and 64 bit state variables */
 
-int32_t iir_df2t(struct iir_state_df2t *iir, int32_t x)
+int32_t iir_process_biquad(int32_t in, int32_t *coef, int64_t *delay)
```
I am wondering if maybe this should be inline to negate the effects of it being called so often

cujomalainey commented on src/include/sof/audio/crossover/crossover.h:
```diff
@@ -0,0 +1,125 @@
+ */
+
+struct lr4_state {
+	int32_t *coef;
```
I think we need to move the individual biquad config/state into a struct so that we can just import that, that way if the ABI changes we will automatically inherit the change.

cujomalainey commented on src/include/sof/audio/crossover/crossover.h:
```diff
@@ -0,0 +1,125 @@
+ */
+struct crossover_state {
+	/* Store the state for each LR4 filter. */
+	lr4_state lowpass[3];
```
magic number

cujomalainey commented on src/include/user/crossover.h:
```diff
@@ -0,0 +1,136 @@
+  *                  o--- LR4 HP0 ---> HIGH sink[1]
+  *
+  *         uint32_t assign_response[channels_in_config]
+  *             -1 = not defined, 0 = use first response, 1 = use 2nd, etc.
```
signed number for unsigned type

cujomalainey commented on src/include/user/crossover.h:
```diff
@@ -0,0 +1,136 @@
+  *         b0 = 1.0, gain = 1.0, and other parameters set to 0
+  *         {0, 0, 0, 0, 1073741824, 0, 16484}
+  */
+struct sof_crossover_config_lr4 {
```
we should be importing this from the iir

sebcarlucci commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,745 @@
+	// align sink streams with their respective configurations
+	list_for_item(clist, &dev->bsink_list) {
+		sink = container_of(clist, struct comp_buffer, source_list);
+		if (sink->sink->state == dev->state) {
```
@cujomalainey If a sink is not ready we may get a gap in the array of sinks, even if the config is valid. Do you know if this scenario is possible?

cujomalainey replied:
I don't know enough how pipelines are constructed or initialized, this is a Liam question.

sebcarlucci replied:
I tried creating this topology (inspired from the demux topologies):
PCM -> Crossover -> DAI
                    |
                    pipeline 2: -> Virtual DAI

And it seems that it is possible that one of the buffers is in a different state from the component.

sebcarlucci committed:
sof: eq_iir: modularize iir_df2t() function

This commit extracts the biquad processing from iir_df2t(). It allows to
reuse the biquad processing code independently of the equalizer
implementation.

Components such as crossover use biquads for processing input.
But it could not do so because iir_df2t() was specific to single output
audio processing.

Signed-off-by: Sebastiano Carlucci <scarlucci@google.com>

sebcarlucci commented:
@cujomalainey previous pr is @ https://github.com/sebcarlucci/sof/tree/crossover/first-pr

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,806 @@
+
+	for (i = 0; i < num_sinks; i++) {
+		/* Get the low pass coefficients */
+		crossover_init_coef_lr4(&coef[j],
```
we are not checking for returned errors here, same for call below

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,806 @@
+ *	       high/low pass filter.
+ * \param[out] lr4 initialized struct
+ */
+static int crossover_init_coef_lr4(struct sof_eq_iir_biquad_df2t *coef,
```
double check where you use this function (and possibly other functions that fail) and make sure you handling the errors because I have found cases where you are not doing that here.

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,806 @@
+	sink_period_bytes = audio_stream_period_bytes(&sink->stream,
+						      dev->frames);
+
+	if (sink->stream.size < config->periods_sink * sink_period_bytes) {
```
shouldnt this be done on a per sink basis?

cujomalainey commented on src/audio/crossover/crossover.c:
```diff
@@ -0,0 +1,806 @@
+	cd->crossover_process(dev, source, sinks, num_sinks, frames);
+
+	/* update components */
+	for (i = 0; i < SOF_CROSSOVER_MAX_STREAMS; i++) {
```
do we not keep the number of sinks on hand? this seems like we are assuming the next element is always null which seems dangerous

sebcarlucci replied:
We keep the number of sinks in the config. But the size of the sink array is initialized to SOF_CROSSOVER_MAX_STREAMS (4).

Maybe a better way to do this would be to initialize the sink array to be cd->config->num_sinks. We can probably safely do that since the config would have already been validated before being assigned. What do you think?

I just had it that way since mux does the same thing too.

sebcarlucci committed:
sof: crossover: Add Crossover component

This commit adds Crossover to the list of SOF components. A crossover
filter can be used to split an input to different frequency bands.
The number of outputs should be set statically in the topology. The user
then uses the control bytes to route the frequency bands to different
outputs. (similar to the demux component).

Signed-off-by: Sebastiano Carlucci <scarlucci@google.com>

sebcarlucci committed:
tools: topology: Add crossover topology files

This commit adds the topology files for the crossover component.
The control bytes are generated by the tools in tune/crossover.

Signed-off-by: Sebastiano Carlucci <scarlucci@google.com>

sebcarlucci committed:
tools: tune: Add tools to generate ctrl bytes for Crossover

This commit adds the tools to generate the control bytes for the
crossover component. To generate the control bytes, run the
example_crossover.m script.

Signed-off-by: Sebastiano Carlucci <scarlucci@google.com>
